src/scripting/game_lua_kernel.cpp

```diff
@@ -5183,15 +5183,16 @@ void game_lua_kernel::mouse_over_hex_callback(const map_location& loc)
 	return;
 }
 
-void game_lua_kernel::select_hex_callback(const map_location& loc)
+void game_lua_kernel::select_hex_callback(map_location& loc)
 {
 	lua_State *L = mState;
 
 	if (!luaW_getglobal(L, "wesnoth", "game_events", "on_mouse_action")) {
 		return;
 	}
+	int ret = lua_gettop(L);
 	lua_push(L, loc.wml_x());
 	lua_push(L, loc.wml_y());
-	luaW_pcall(L, 2, 0, false);
-	return;
+	luaW_pcall(L, 2, 2, false);
+	luaW_tolocation(L, ret, loc);
```

Comment:
What is expected to have happened here - is the on_mouse_action event expected to modify the location in UMC WML / Lua code?

Reply:
What would be the usecase for that? I'm not seeing a reason why the game would have a better idea of where the user intended to click than the user themselves.

Reply:
Once the callback returns, the game will proceed with the standard effect of clicking the specified hex. By returning an invalid hex (eg -9,-9), this can be prevented.

As for returning a valid hex other than the one clicked, I can't think of anything concrete right now. I think there _is_ use-case, though – if there was some mechanic where two hexes are "paired", it might make sense to have a handler where you click one of them but the action is instead performed on the other.

Reply:
Having it say that it's handled the click makes sense to me.

However, I don't see the purpose of "paired" hexes instead of using the existing options to select units and hexes.
Generally, I feel it would better to avoid having an extra edge-case to support in the engine.

Reply:
Returning a different location would simulate an actual click on that location, so the action would depend on what's on that location and the unit (if any) that's currently selected. I'm not quite sure whether a "paired" hexes mechanic would require an actual simulated click, though.

Reply:
But what UX does that enable? The Lua code can easily find out what's on that other hex, it presumably chooses that other hex based on something that the Lua code knows about the scenario, so why doesn't the Lua just call whatever code it wants to directly?

The answers that you've given so far read to me as "we can add more complexity" without saying what benefits come from that complexity. When there are possible benefits but no concrete example of what would use them, I'd suggest putting a "decide whether to do this" task on the roadmap, along with encouraging the UMC authors to look at the roadmap and comment if they would use the potential features.